FreeRTOS_IP.c

```diff
@@ -3495,159 +3495,117 @@ uint16_t usGenerateProtocolChecksum( const uint8_t * const pucEthernetBuffer,
-    }
-
-    /* Word (32-bit) aligned, do the most part. */
-    xLastSource.u32ptr = ( xSource.u32ptr + ( uxDataLengthBytes / 4U ) ) - 3U;
-
-    /* In this loop, four 32-bit additions will be done, in total 16 bytes.
-     * Indexing with constants (0,1,2,3) gives faster code than using
-     * post-increments. */
-    while( xSource.u32ptr < xLastSource.u32ptr )
-    {
-        /* Use a secondary Sum2, just to see if the addition produced an
-         * overflow. */
-        xSum2.u32 = xSum.u32 + xSource.u32ptr[ 0 ];
+        /* Transform a pointer to a numeric value, in order to determine
+         * the alignment. */
+        uxBufferAddress = ( uintptr_t ) pucData;
 
-        if( xSum2.u32 < xSum.u32 )
+        if( ( uxBufferAddress & 1U ) != 0U )
         {
-            ulCarry++;
+            ulAccum = ( ( ulAccum & 0xffU ) << 8 ) | ( ( ulAccum & 0xff00U ) >> 8 );
+            usTerm = pucData[ 0 ];
+            usTerm = FreeRTOS_htons( usTerm );
+            /* Now make pucData 16-bit aligned. */
+            uxBufferAddress++;
+            /* One byte has been summed. */
+            uxBytesLeft--;
+            xUnaligned = pdTRUE;
         }
 
-        /* Now add the secondary sum to the major sum, and remember if there was
-         * a carry. */
-        xSum.u32 = xSum2.u32 + xSource.u32ptr[ 1 ];
+        /* The alignment of 'pucData' has just been tested and corrected
+         * when necessary.
+         */
+        pusPointer = ( const uint16_t * ) uxBufferAddress;
 
-        if( xSum2.u32 > xSum.u32 )
+        /* Sum 'uxUnrollCount' shorts in each loop. */
+        while( uxBytesLeft >= ( sizeof( *pusPointer ) * uxUnrollCount ) )
         {
-            ulCarry++;
+            ulAccum += pusPointer[ 0 ];
+            ulAccum += pusPointer[ 1 ];
+            ulAccum += pusPointer[ 2 ];
+            ulAccum += pusPointer[ 3 ];
+            ulAccum += pusPointer[ 4 ];
+            ulAccum += pusPointer[ 5 ];
+            ulAccum += pusPointer[ 6 ];
+            ulAccum += pusPointer[ 7 ];
+            ulAccum += pusPointer[ 8 ];
+            ulAccum += pusPointer[ 9 ];
+            ulAccum += pusPointer[ 10 ];
+            ulAccum += pusPointer[ 11 ];
+            ulAccum += pusPointer[ 12 ];
+            ulAccum += pusPointer[ 13 ];
+            ulAccum += pusPointer[ 14 ];
+            ulAccum += pusPointer[ 15 ];
+
+            uxBytesLeft -= sizeof( *pusPointer ) * uxUnrollCount;
+            pusPointer = &( pusPointer[ uxUnrollCount ] );
         }
 
-        /* And do the same trick once again for indexes 2 and 3 */
-        xSum2.u32 = xSum.u32 + xSource.u32ptr[ 2 ];
-
-        if( xSum2.u32 < xSum.u32 )
+        /* Between 0 and 7 shorts might be left. */
+        while( uxBytesLeft >= sizeof( *pusPointer ) )
         {
-            ulCarry++;
+            ulAccum += *pusPointer;
+            uxBytesLeft -= sizeof( *pusPointer );
+            pusPointer = &( pusPointer[ 1 ] );
```

Comment:
see previous comment on the ++ operator.

Reply:
I didn't use the indirect post increment addressing mode, because I feared that the compiler would need 3 instructions: load, add, and increment, times 16.

But with optimisation enabled, the compiler only increments `pusPointer` with 32 at the end of each loop. Thus giving the same performance.

It certainly looks better than:
~~~c
pusPointer = &( pusPointer[ 1 ] );
~~~
Although Coverity did not complain about this type of assignment.

Thanks, Hein

Reply:
Depending upon how the compiler optimizes I would expect pusPointer to stay in a register for this function.  Then that addressing mode will leave the loop with pusPointer updated as a "side effect" of the single instruction.  I am doing a test.